### ISSUE TITLE
Allow type converters to be async

### ISSUE THREAD

**Issue by clragon** [simolus3/drift#1895]
**Is your feature request related to a problem? Please describe.**
When querying alot of data, JSON decode can take a toll on the UI running in the main isolate.

**Describe the solution you'd like**
Changing type converters to allow returning a `FutureOr<T>` could be a non-breaking change
while enabling the usage of dedicated JSON decoding isolates inside of the converter.

Manually decoding the JSON after the query would mean we have to introduce extra code in the query functions,
as well as a new type that allows turning the already converted data from the database into a new type that now holds the json decoded data. This would take us further away from the ease of generated types and readable query functions.

This could be a possibly easier to implement alternative (if feasible) to #1872.

**Comment by simolus3** [comment]
>  Changing type converters to allow returning a `FutureOr<T>` could be a non-breaking change

I like this idea, but it would probably still be breaking since we currently generate a synchronous factory to map rows to data classes (`DataClass.fromData`). That would have to become async (or at least a `FutureOr` but then we may as well make it a `Future` since users need to await it either way).

Perhaps there's a way to preserve backwards-compatibility here by only generating an async mapping method if an asynchronous type converter is used :thinking: But then we may need a special interface for those to detect their use.

**Comment by simolus3** [comment]
Now I wonder how painful it would be to write converters as `extends TypeConverter<Future<YourType>, String>` and let drift generate a future in your data classes which you would have to await? It certainly introduces an additional bit of boilerplate, but perhaps it's an acceptable solution if you only need it for a few columns that are storing large json strings?

**Comment by clragon** [comment]
In my eyes, that would defeat the point entirely.
The reason we wanted asynchronous type converters was so that we can convert json inside of isolates.
Since converting it inside the drift isolates doesnt seem like a thing you would like to implement, it would be nice if we can do it ourselves in the type converters.

Of course we can already do this ourselves by just:
- query our data normally, no json conversion with a type that uses strings
- converts the json inside of the query function with an isolate
- transfers all data from the original query type to a new type that is identical but has a datatype instead of a string field

the problem is just that this makes us write every single object twice and we have then arrived at a point where we are writing tons of boilerplate code.

if we replace the string with a future on our data objects, we still would have to transfer all the data to a new type so that in the end we und up with a normal data object.

**Comment by simolus3** [comment]
That's a fair point, thanks for the feedback.

I've now spent a fair bit of time trying to implement async type converters, but it just appeared to be a bottomless bit of more stuff breaking the further I went:

1. Given that it may call async converters, `ResultSetImplementation.map` (the method going from a row as a map to a Dart object) would have to be asynchronous.
2. For the other direction, `Insertable.toColumns()` would have to be asynchronous too.
3. There are `JsonTypeConverter`s which apply the conversion to json as well, so all of drift's builitn json serialization would have to become asynchronous as well. Nowadays I'm not too happy with json serialization being builtin to drift in the first place, but it's there now and I don't think it can really stop being synchronous.
4. On columns that have a type converter applied, the `equalsValue` method can be used to generate an `==` in SQL based on the mapped value. This would have to start being asynchronous too, which means that the _entire_ query builder API would have to be async.

I'm ok with 1 and 2, but 3 and 4 are really causing problems and I haven't found a decent solution for them unfortunately.

One other approach that might work: Drift supports [custom row classes](https://drift.simonbinder.eu/docs/advanced-features/custom_row_classes/), what if we allowed the mapping there to be asynchronous? For instance, you could then write:

```dart
class MyRowClass {
  // bunch of simple fields
  final MyComplexJsonObject json;

  static Future<MyRowClass> mapFromDb({
    /* other fields */,
    String json,
  }) async {
    // use `compute()` or another approach to decode json
   }
}

@UseDataClass(MyRowClass, constructor: 'mapFromDb')
class MyTable extends Table {
  TextColumn get json => text()(); // no type converter here, mapping done while loading
}
```

This solves the pain point of having to do two awaits, since drift would await the `mapFromDb` call internally after running a query. However, you'd have to write your own data class if it uses the large JSON field. Given that this should hopefully only be needed in few cases, do you think that could work? It's a tradeoff between the complexity of doing this in general and some convenience.

**Comment by clragon** [comment]
unfortunately this leads to the exact same problem mentioned above:
we have to write everything twice.
but I suppose this solution would be better integrated at least.
the problems youve mentioned seem pretty severe.
I think we would be okay with this as a solution.

**Comment by simolus3** [comment]
Yeah, I agree it could ideally be more convenient to use but that's the tradeoff one has to make.

Support for async mappings to custom row classes is implemented in b9a605ed2510eacb6d53c47d6698a6720faa0c78. Let's see how that works out, if you have follow-up ideas based on that let's open a new issue.
